### PR TITLE
Lua, SQL: Implement Zipacna pathing and interaction with Cermet Doors

### PR DESCRIPTION
**Pull request by hooksta4**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Currently, Zipacna only spawns in one place and paths through the doors regardless if they are open or closed. 
This PR addresses the following: (I made no modifications to Zipacna's abilities or stats)
- adds the 4 spawn points located as shown here. 
image
image
image
https://www.youtube.com/watch?v=nWjGMWHsg9k

The doors are labelled as such (from west side to east side)
Basement Loop - |1st Blue Door | Middle Blue Door | Middle Yellow Door | 1st Yellow Door| - Basement loop

- There are 4 doors controlled by a Yellow or Blue Monolith. When one is activated the other "side doors" are closed. 
** *Note - It is possible that Zipacna can get stuck between the middle yellow and first yellow door as well as the middle blue door and first blue door. If someone switches while he is pathing between**

Zipacna's behavior is to path to a door. If the door is closed, it will turn around and go to the next door in the sequence.
Example: Middle Yellow door was closed, Zipacana will turn around and go to the middle blue door. If the middle blue door is open, Zipacna will proceed to the 1st Blue door, make a loop in the basement and then return to the first blue door. If however, the yellow switch is then triggered, the middle blue door is now closed and once Zipacna arrives will again turn around and go to the middle yellow door. 

Due to there needing to be checks on both sides of the door, vars were used to track the last door Zip passed through. It is hard to follow but the video shows an hour and a half of Zipacna pathing with all the scenarios I could come up with that could happen. https://youtu.be/rWGuV6pqJww

TODO: I believe it is more of a navmesh thing but if he is engaged and the players wipe/zone Zipacna will likely path through walls back to its spawn point, at which point, will begin his pathing again. So suck less?

## Steps to test these changes
!spawnmob 17502570 (Will spawn in 1 of 4 locations.
!gotoid 17502570 - follow Zipacna and you will observe it path to each sides respective first doors. At that point, the pathing between doors takes over.

I understand that there is most likely not a way to clean up the path points, but if there is a way to clean up the door logic, I am open to suggestions.

### CODE REVIEW

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (Xaver-DaRed) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:337]: Shouldnt you be using `elseif` for this ones instead of `if`s?

Also, there are only 4 options where it can spawn, all of them with severely diferent X positions. Can you check if this keeps working if you simplify the checks to only check for the X position?

Could also edit the X check to be more generous in case it gets slightly off.

Also, you have to allign comments properly, and remove the prints.

Comment (Xaver-DaRed) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:337]: End result would look like this
```lua
    -- Zip popped on blue short side (!pos -196, 0, 396)
    if pos.x <= -195 then
        mob:pathThrough(blueShortSideSpawnPath, xi.path.flag.COORDS)

    -- Zip popped at the blue basement spawn point (!pos -121, 16, 420)
    elseif pos.x <= -120 then
        mob:pathThrough(blueBasementSpawnPath, xi.path.flag.COORDS)

    -- Zip popped at the Yellow Short Side Spawn Point - going to first yellow door (!pos 191, 0, 399)
    elseif pos.x >= 190 then
        mob:pathThrough(yellowShortSideSpawnPath, xi.path.flag.COORDS)

    -- Zip popped in the yellow side basement - going to first yellow door
    elseif pos.x >= 140 then
        mob:pathThrough(yellowBasementSpawnPath, xi.path.flag.COORDS)
    end
```

Comment (hooksta4) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:337]: Ill give it a go - I think we can get away without the Z check but the Y will be needed I would imagine.

Comment (hooksta4) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:337]: Looks like I was wrong - seems to be just fine with the x -

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:362]: I'm struggling to understand what you're checking here?

You check the distance to the door (you can be using `mob:checkDistance(firstBlueDoor) <= 5`), and then you're checking some other point? Can you extract that to something named? Is it a part of one of the paths?

Then you're choosing which path to do depending on whether or not the door is closed. Fine, but the "if path is closed" is one other other other paths in reverse, no? I asked you in ASB or HXI discord to look at the `xi.path.flag.REVERSE` flat that `Guivre` uses so you don't have to have so much path data.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:396]: You repeat this exact pattern 4 times using different points and paths, extract it out into a function or a table, etc.

File: sql/nm_spawn_points.sql
Comment (zach2good) [inline sql/nm_spawn_points.sql:5666]: I don't know how this system works, will it just pick randomly from the points it has listed?

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:322]: It would be more obvious if you specify the spawn points as x/y/z tables, with a comment saying this is what's listed in the nm spawn points SQL, and then do `checkDistance()` on each of them

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:299]: The naming of these is inconistant, you're mixing up "position" and "colour" in the naming of these. This is also presumably just the reverse of `firstBluetoBlueMiddle`. See my comment somewhere below about using the reverse flag.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:97]: Comment formatting:
```lua
-- From the first blue door to the blue basement loop
local firstBlueToBasement =
{
    { x = -162, y = 0, z = 380 }, -- First Blue Door
    { x = -171, y = 0, z = 379 },
```

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:12]: Do you need these, if you're looking up the door entities themselves? You can just use `getPos()` on them, and `checkDistance()` for comparisons

**Review by zach2good — CHANGES_REQUESTED** [review]

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:473]: Instead of setting these, would it work to use `not entity:isFollowingPath()` to make sure you aren't constantly resetting the path with `pathThrough()`?

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (hooksta4) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:362]: If I am understanding your question correctly, there needs to be a check on both sides of the door. The reverse paths can work for the path between middle doors to the first doors but will not for the basement routes. I did not include them as I thought it would be confusing. Regardless, I feel all the paths would still need to be there due to coming from one side or the other. 

Maybe my approach is not the best way for this - thinking on it, I might be able to first check the status of the door (open/closed) and then determine which side of the door Zipacna is on. And from there, path accordingly.

File: sql/nm_spawn_points.sql
Comment (hooksta4) [reply sql/nm_spawn_points.sql:5666]: That is my understanding, yes. Prior to having `INSERT INTO `nm_spawn_points` VALUES (17502570,0,-196.187,0.51,396.813); -- Zipacna` and
`UpdateNMSpawnPoint(mob:getID())` - it would spawn at the same point that was listed in mob_spawn_points.sql - now it randomly picks between the 4 locations.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (hooksta4) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:299]: Indeed, the reverse will not work for the basement loops as Zip goes in the same direction regardless. The reverse flag would have it go counter to that. For the sake of uniformity, I opted to not include those for the middle doors as I thought it might be confusing.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (Frankie-hz) [inline scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:444]: Remove prints

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (zach2good) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:362]: Do you even need to know which side of the door you're on? You already know which path you're using so if you're within whatever range of the door and it's closed, you know to reverse the current path. The side of approach is irrelevant because you already know which path you're using. If you don't, you can put your paths in a table and keep track of the index as a local var

**Comment by zach2good** [comment]
> TODO: I believe it is more of a navmesh thing but if he is engaged and the players wipe/zone Zipacna will likely path through walls back to its spawn point, at which point, will begin his pathing again. So suck less?

This isn't a navmesh issue. This is a pathing issue. This would be a very good opportunity for you to write a helper function that takes a current patrol path, finds the nearest point to a given entity on that path, and then uses that slice of the patrol path to call to `pathThrough`. Essentially a "resume this path from the closest point" - rather than returning to spawn or the very start of the path

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:337]: Helping Hook out with this - So will comment on some of this on his behalf.
Removed all of these `pos.x` checks in lieu of `checkDistance` calls using a table for the spawn points now listed at top of file.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:396]: Helping Hook out with this - So will comment on some of this on his behalf.
Extracted these repeated calls into a helper function that determines the path finding logic.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:362]: Helping Hook out with this - So will comment on some of this on his behalf.
Completely changed the door logic and used a local var (scoped to the lua) to determine the direction Zip is going.
Also updated to use `checkDistance` calls and `xi.path.flag.REVERSE` and shortened his pathing node list.
The pathing node list has also been converted to a table.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:322]: Helping Hook out with this - So will comment on some of this on his behalf.
This has been extracted into a table at the start of the file.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:299]: Helping Hook out with this - So will comment on some of this on his behalf.
With the removal of excess local vars and using an indexed table to determine pathing, this should resolve most of this I believe.  I am still not a large fan of the naming convention myself, but I kept things as close to his as I could.

If he needs to do extra modifications on names, I will leave that to him for now.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:97]: Helping Hook out with this - So will comment on some of this on his behalf.
These should all be fixed now I believe.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:12]: Helping Hook out with this - So will comment on some of this on his behalf.
Swapped these to `GetNPCByID` and then using `getPos()` to determine their locations instead of hard-coding them

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:473]: Helping Hook out with this - So will comment on some of this on his behalf.
Since these have been extracted to a helper function, and `pathThrough()` is only getting called when in the presence of a door now, I am not sure that we need to use `isFollowingPath()` as the paths now end at each door where the check is performed to find a new path.

Might need to check the usage in the updated ammended commit to see if there is still a need to swap in some `isFollowingPath()` logic.

File: scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua
Comment (0x05010705) [reply scripts/zones/VeLugannon_Palace/mobs/Zipacna.lua:444]: Helping Hook out with this - So will comment on some of this on his behalf.
These have been removed.

**Comment by 0x05010705** [comment]
Well, went to amend his commit, but I don't have Push Access to his fork - So I will send the updated changes to @hooksta4 and have him amend this PR from his end.

**Comment by 0x05010705** [comment]
This PR can be closed - Was not able to get the changes amended to this fork/branch.
This was re-opened as a new PR with the requested changes that I can continue to make adjustments on if needed.
New PR: https://github.com/LandSandBoat/server/pull/6516

**Comment by WinterSolstice8** [comment]
as directed, superceded by #6516

**Pull request closed without merging**
